### PR TITLE
Added OAuth feature for http sink plugin.

### PR DESCRIPTION
**Pull request by Nikitapaliwal123**
Added OAuth feature for http sink plugin.
https://b.corp.google.com/issues/271967808

### CODE REVIEW

**Comment by MrRahulSharma** [comment]
Add the JIRA link in the description and the PR description.

File: src/main/java/io/cdap/plugin/http/sink/batch/HTTPRecordWriter.java [hunk 1]
```diff
@@ -103,7 +110,27 @@ private void executeHTTPService() throws IOException {
...
     do {
       HttpURLConnection conn = null;
-      Map<String, String> headers = config.getRequestHeadersMap();
+
+      Map<String, String> headers = null;
+
+      // auth check
+      AuthType authType = config.getAuthType();
+        ArrayList<Header> clientHeaders = new ArrayList<>();
+        switch (authType) {
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/http/sink/batch/HTTPRecordWriter.java]: Nitpik: Correct the indentation.

File: src/main/java/io/cdap/plugin/http/sink/batch/HTTPRecordWriter.java [hunk 2]
```diff
@@ -103,7 +110,27 @@ private void executeHTTPService() throws IOException {
     IOException exception = null;
     do {
       HttpURLConnection conn = null;
-      Map<String, String> headers = config.getRequestHeadersMap();
+
+      Map<String, String> headers = null;
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/http/sink/batch/HTTPRecordWriter.java]: The headers can remain null in case the authType does not match the OAUTH2 or SERVICE_ACCOUNT which is leading to null pointer exception on line#149.

File: src/main/java/io/cdap/plugin/http/sink/batch/HTTPSinkConfig.java [hunk 3]
```diff
@@ -24,16 +24,22 @@
 import io.cdap.cdap.api.annotation.Name;
 import io.cdap.cdap.api.data.schema.Schema;
 import io.cdap.cdap.etl.api.FailureCollector;
+import io.cdap.cdap.etl.api.validation.InvalidConfigPropertyException;
 import io.cdap.cdap.format.StructuredRecordStringConverter;
 import io.cdap.plugin.common.ReferenceNames;
 import io.cdap.plugin.common.ReferencePluginConfig;
 
+import io.cdap.plugin.http.source.batch.HttpBatchSourceConfig;
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/http/sink/batch/HTTPSinkConfig.java]: Remove Unused imports.

File: src/main/java/io/cdap/plugin/http/sink/batch/HTTPSinkConfig.java [hunk 4]
```diff
@@ -24,16 +24,22 @@
...
+import io.cdap.plugin.http.source.common.http.AuthType;
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/http/sink/batch/HTTPSinkConfig.java]: It seems like the common classes are present in the Http source package, Can you refactor those into a common package ?

Comment (Nikitapaliwal123) [reply src/main/java/io/cdap/plugin/http/sink/batch/HTTPSinkConfig.java]: .* imports are not allowed

File: src/main/java/io/cdap/plugin/http/common/BaseHttpConfig.java [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+import java.io.File;
+import java.util.ArrayList;
+import java.util.Optional;
+import javax.annotation.Nullable;
+import org.apache.http.Header;
+
+/**
+ *
+ */
+public class BaseHttpConfig extends PluginConfig {
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/http/common/BaseHttpConfig.java]: 1. This class should be an abstract class.
2. This class should be extending from the ReferencePluginConfig.

File: src/main/java/io/cdap/plugin/http/common/BaseHttpConfig.java [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+import io.cdap.plugin.http.source.common.BaseHttpSourceConfig;
+import java.io.File;
+import java.util.ArrayList;
+import java.util.Optional;
+import javax.annotation.Nullable;
+import org.apache.http.Header;
+
+/**
+ *
+ */
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/http/common/BaseHttpConfig.java]: Please add a descriptive class details.

File: src/main/java/io/cdap/plugin/http/common/BaseHttpConfig.java [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+            if (!Optional.ofNullable(getServiceAccountJson()).isPresent()) {
+                collector.addFailure("Service Account JSON can not be empty.",
+                                "Please provide Service Account JSON.")
+                        .withConfigProperty(PROPERTY_NAME_SERVICE_ACCOUNT_JSON);
+            }
+        }
+        return collector.getValidationFailures().size() == 0;
+    }
+
+     public void validateOAuth(FailureCollector collector) {
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/http/common/BaseHttpConfig.java]: Use the validate method to validate all the local members of this class and avoid writing specific validate methods here.

File: src/main/java/io/cdap/plugin/http/common/http/HttpClient.java [hunk 8]
```diff
@@ -144,7 +144,9 @@ private CloseableHttpClient createHttpClient() throws IOException {
         break;
       case SERVICE_ACCOUNT:
         // get accessToken from service account
-        accessToken = OAuthUtil.getAccessTokenByServiceAccount(config);
+        accessToken = OAuthUtil.getAccessToken(config.getServiceAccountScope(), config.isServiceAccountJson(),
+                                               config.getServiceAccountJson(), config.isServiceAccountFilePath(),
+                                               config.getServiceAccountFilePath());
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/http/common/http/HttpClient.java]: Revert this change, this is not required.

File: src/main/java/io/cdap/plugin/http/common/http/OAuthUtil.java [hunk 9]
```diff
@@ -67,29 +67,31 @@ public static String getAccessTokenByRefreshToken(CloseableHttpClient httpclient
...
-          .createScoped(scopeSet);
+                .createScoped(scopeSet);
+      } else if (isServiceAccountFilePath && !Strings.isNullOrEmpty(serviceAccountFilePath)
+              && !BaseHttpSourceConfig.PROPERTY_AUTO_DETECT_VALUE.equals(serviceAccountFilePath)) {
+        credential = GoogleCredentials.fromStream(new FileInputStream(serviceAccountFilePath))
+                .createScoped(scopeSet);
       } else {
         credential = GoogleCredentials.getApplicationDefault()
-          .createScoped(scopeSet);
+                .createScoped(scopeSet);
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/http/common/http/OAuthUtil.java]: These changes are not required.

File: src/main/java/io/cdap/plugin/http/sink/batch/HTTPRecordWriter.java [hunk 10]
```diff
@@ -103,7 +110,32 @@ private void executeHTTPService() throws IOException {
...
+            accessToken = OAuthUtil.getAccessToken(config.getServiceAccountScope(), config.isServiceAccountJson(),
+                                                   config.getServiceAccountJson(), config.isServiceAccountFilePath(),
+                                                   config.getServiceAccountFilePath());
+            clientHeaders.add(new BasicHeader("Authorization", "Bearer " + accessToken));
+            headers = config.getHeadersMap(String.valueOf(clientHeaders));
+            break;
+        }
+      if (headers.equals(null))  {
+        headers = config.getRequestHeadersMap();
+      }
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/http/sink/batch/HTTPRecordWriter.java]: This code segment should be shared from the HttpClient class.

File: src/main/java/io/cdap/plugin/http/source/common/BaseHttpSourceConfig.java [hunk 11]
```diff
@@ -482,24 +379,10 @@ public Boolean getCsvSkipFirstRow() {
     return Boolean.parseBoolean(csvSkipFirstRow);
   }
 
-  public AuthType getAuthType() {
-    return AuthType.fromValue(authType);
-  }
-
   public void setAuthType(String authType) {
     this.authType = authType;
   }
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/http/source/common/BaseHttpSourceConfig.java]: This should move to parent class.

File: src/main/java/io/cdap/plugin/http/sink/batch/HTTPSinkConfig.java [hunk 12]
```diff
@@ -448,6 +462,16 @@ public Builder setFailOnNon200Response(Boolean failOnNon200Response) {
...
+    public Builder setOauth2Enabled(String oauth2Enabled) {
+      this.oauth2Enabled = oauth2Enabled;
+      return this;
+    }
+
+    public Builder setAuthType(String authType) {
+      this.authType = authType;
+      return this;
+    }
+
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/http/sink/batch/HTTPSinkConfig.java]: These methods are not required.

File: src/main/java/io/cdap/plugin/http/source/batch/HttpBatchSourceConfig.java [hunk 13]
```diff
@@ -22,8 +22,8 @@
  * Provides all the configurations required for configuring the {@link HttpBatchSource} plugin.
  */
 public class HttpBatchSourceConfig extends BaseHttpSourceConfig {
-  protected HttpBatchSourceConfig(String referenceName) {
-    super(referenceName);
+  protected HttpBatchSourceConfig(String referenceName, String authType, String oauth2Enabled) {
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/http/source/batch/HttpBatchSourceConfig.java:25]: Write a builder in the BaseHttpConfig, which is the parent of the HttpBaseSourceConfigBuilder to avoid this change.

File: src/main/java/io/cdap/plugin/http/source/common/BaseHttpSourceConfig.java [hunk 14]
```diff
@@ -880,110 +709,14 @@ PAGINATION_INDEX_PLACEHOLDER, getPaginationType()),
...
-        String reasonBasicAuth = "Basic Authentication is enabled";
-        if (!containsMacro(PROPERTY_USERNAME)) {
-          assertIsSet(getUsername(), PROPERTY_USERNAME, reasonBasicAuth);
-        }
-        if (!containsMacro(PROPERTY_PASSWORD)) {
-          assertIsSet(getPassword(), PROPERTY_PASSWORD, reasonBasicAuth);
-        }
-        break;
-    }
+    validateOAuth(failureCollector);
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/http/source/common/BaseHttpSourceConfig.java]: Implement validate method in the parent class instead, and move the logic there, You can call super.validate from this method instead.

File: src/main/java/io/cdap/plugin/http/source/streaming/HttpStreamingSourceConfig.java [hunk 15]
```diff
@@ -36,8 +36,8 @@ public class HttpStreamingSourceConfig extends BaseHttpSourceConfig {
   @Macro
   protected Long maxPagesPerFetch;
 
-  protected HttpStreamingSourceConfig(String referenceName) {
-    super(referenceName);
+  protected HttpStreamingSourceConfig(String referenceName, String authType, String oauth2Enabled) {
+    super(referenceName, authType, oauth2Enabled);
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/http/source/streaming/HttpStreamingSourceConfig.java]: This change is not correct and it should be resolved by the correct builder class creation in the parent class.

File: src/test/java/io/cdap/plugin/http/source/common/pagination/page/JSONUtilTest.java
Comment (MrRahulSharma) [inline src/test/java/io/cdap/plugin/http/source/common/pagination/page/JSONUtilTest.java:1]: This class should also change its package from 'package io.cdap.plugin.http.source.common.pagination.page' to 'package io.cdap.plugin.http.common.pagination.page'

File: src/test/java/io/cdap/plugin/http/source/common/pagination/page/JSONUtilTest.java [hunk 16]
```diff
@@ -16,6 +16,7 @@
 package io.cdap.plugin.http.source.common.pagination.page;
 
 import com.google.gson.JsonObject;
+import io.cdap.plugin.http.common.pagination.page.JSONUtil;
```
Comment (MrRahulSharma) [inline src/test/java/io/cdap/plugin/http/source/common/pagination/page/JSONUtilTest.java]: This change is not required.

File: src/main/java/io/cdap/plugin/http/common/BaseHttpConfig.java [hunk 17]
```diff
@@ -0,0 +1,346 @@
+package io.cdap.plugin.http.common;
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/http/common/BaseHttpConfig.java:17]: Please add the license doc here.

File: src/main/java/io/cdap/plugin/http/common/http/HttpClient.java [hunk 18]
```diff
@@ -125,6 +126,20 @@ private CloseableHttpClient createHttpClient() throws IOException {
...
+      for (Map.Entry<String, String> headerEntry : this.headers.entrySet())  {
+        clientHeaders.add(new BasicHeader(headerEntry.getKey(), headerEntry.getValue()));
+      }
+    }
+    httpClientBuilder.setDefaultHeaders(clientHeaders);
+
+    return httpClientBuilder.build();
+  }
+
+  public static ArrayList<Header> getClientHeaders(BaseHttpConfig config) throws IOException {
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/http/common/http/HttpClient.java:142]: Please add function doc.

File: src/main/java/io/cdap/plugin/http/common/http/HttpClient.java [hunk 19]
```diff
@@ -148,16 +163,7 @@ private CloseableHttpClient createHttpClient() throws IOException {
         clientHeaders.add(new BasicHeader("Authorization", "Bearer " + accessToken));
         break;
     }
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/http/common/http/HttpClient.java:165]: This should ideally be inside the config object.

File: src/main/java/io/cdap/plugin/http/common/http/OAuthUtil.java [hunk 20]
```diff
@@ -67,7 +67,10 @@ public static String getAccessTokenByRefreshToken(CloseableHttpClient httpclient
     return jsonElement.getAsString();
   }
 
-  public static String getAccessTokenByServiceAccount(BaseHttpSourceConfig config) throws IOException {
+  public static String getAccessTokenByServiceAccount(BaseHttpConfig config) throws IOException {
+    return getAccessToken(config);
+  }
+  public static String getAccessToken(BaseHttpConfig config) throws IOException {
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/http/common/http/OAuthUtil.java]: This is not required now.

File: src/main/java/io/cdap/plugin/http/common/http/OAuthUtil.java [hunk 21]
```diff
@@ -80,16 +83,16 @@ public static String getAccessTokenByServiceAccount(BaseHttpSourceConfig config)
...
         credential = GoogleCredentials.fromStream(new FileInputStream(config.getServiceAccountFilePath()))
-          .createScoped(scopeSet);
+                .createScoped(scopeSet);
       } else {
         credential = GoogleCredentials.getApplicationDefault()
-          .createScoped(scopeSet);
+                .createScoped(scopeSet);
       }
       accessToken = credential.refreshAccessToken().getTokenValue();
     } catch (Exception e) {
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/http/common/http/OAuthUtil.java:95]: Spacing changes are not required in this file.

File: src/main/java/io/cdap/plugin/http/common/BaseHttpConfig.java [hunk 22]
```diff
@@ -0,0 +1,394 @@
...
+                    assertIsSet(getUsername(), PROPERTY_USERNAME, reasonBasicAuth);
+                }
+                if (!containsMacro(PROPERTY_PASSWORD)) {
+                    assertIsSet(getPassword(), PROPERTY_PASSWORD, reasonBasicAuth);
+                }
+                break;
+        }
+    }
+
+    public static ArrayList<Header> getAuthorizationHeaders(BaseHttpConfig config) throws IOException {
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/http/common/BaseHttpConfig.java]:
```suggestion
    public Header getAuthorizationHeader() throws IOException {
```

File: src/main/java/io/cdap/plugin/http/common/BaseHttpConfig.java [hunk 23]
```diff
@@ -0,0 +1,394 @@
...
+        ArrayList<Header> clientHeaders = new ArrayList<>();
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/http/common/BaseHttpConfig.java]:
```suggestion
```

File: src/main/java/io/cdap/plugin/http/common/BaseHttpConfig.java [hunk 24]
```diff
@@ -0,0 +1,394 @@
...
+                        config.getRefreshToken());
+                clientHeaders.add(new BasicHeader("Authorization", "Bearer " + accessToken));
+                break;
+            case SERVICE_ACCOUNT:
+                // get accessToken from service account
+                accessToken = OAuthUtil.getAccessTokenByServiceAccount(config);
+                clientHeaders.add(new BasicHeader("Authorization", "Bearer " + accessToken));
+                break;
+        }
+        return clientHeaders;
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/http/common/BaseHttpConfig.java]:
```suggestion
        return null;
```

File: src/main/java/io/cdap/plugin/http/common/http/HttpClient.java [hunk 25]
```diff
@@ -125,29 +126,7 @@ private CloseableHttpClient createHttpClient() throws IOException {
...
-                                                                    config.getRefreshToken());
-        clientHeaders.add(new BasicHeader("Authorization", "Bearer " + accessToken));
-        break;
-      case SERVICE_ACCOUNT:
-        // get accessToken from service account
-        accessToken = OAuthUtil.getAccessTokenByServiceAccount(config);
-        clientHeaders.add(new BasicHeader("Authorization", "Bearer " + accessToken));
-        break;
-    }
+    ArrayList<Header> clientHeaders = BaseHttpConfig.getAuthorizationHeaders(config);
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/http/common/http/HttpClient.java]:
```suggestion
    ArrayList<Header> clientHeaders = new ArrayList<>();
    Header authHeader = config.getAuthorizationHeader();
```

File: src/main/java/io/cdap/plugin/http/sink/batch/HTTPRecordWriter.java [hunk 26]
```diff
@@ -103,7 +108,14 @@ private void executeHTTPService() throws IOException {
...
       HttpURLConnection conn = null;
-      Map<String, String> headers = config.getRequestHeadersMap();
+
+      Map<String, String> headers = null;
+
+      ArrayList<Header> clientHeaders = BaseHttpConfig.getAuthorizationHeaders(config);
+      headers = config.getHeadersMap(String.valueOf(clientHeaders));
+      if (headers.equals(null))  {
+        headers = config.getRequestHeadersMap();
+      }
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/http/sink/batch/HTTPRecordWriter.java]:
```suggestion
      Map<String, String> headers = config.getRequestHeadersMap();
      Header authHeader = config.getAuthorizationHeader();
      if (authHeader != null)  {
      headers.putAll(config.getMapFromKeyValueString(String.valueOf(clientHeaders)));
      }
```

File: src/main/java/io/cdap/plugin/http/common/http/HttpClient.java [hunk 27]
```diff
@@ -126,7 +126,13 @@ private CloseableHttpClient createHttpClient() throws IOException {
     }
     httpClientBuilder.setDefaultCredentialsProvider(credentialsProvider);
 
-    ArrayList<Header> clientHeaders = BaseHttpConfig.getAuthorizationHeaders(config);
+    ArrayList<Header> clientHeaders = new ArrayList<>();
+
+    Header authorizationHeader = config.getAuthorizationHeaders();
+
+    if (authorizationHeader != null) {
+         clientHeaders.add(authorizationHeader);
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/http/common/http/HttpClient.java]: Nit: Correct Indentation.

File: src/main/java/io/cdap/plugin/http/common/BaseHttpConfig.java [hunk 28]
```diff
@@ -358,31 +358,28 @@ public void validate(FailureCollector failureCollector) {
         }
     }
 
-    public static ArrayList<Header> getAuthorizationHeaders(BaseHttpConfig config) throws IOException {
-        ArrayList<Header> clientHeaders = new ArrayList<>();
+    public Header getAuthorizationHeaders() throws IOException {
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/http/common/BaseHttpConfig.java]: Nit : Should be getAuthorizationHeader

File: src/main/java/io/cdap/plugin/http/sink/batch/HTTPRecordWriter.java [hunk 29]
```diff
@@ -109,13 +108,11 @@ private void executeHTTPService() throws IOException {
     do {
       HttpURLConnection conn = null;
 
-      Map<String, String> headers = null;
+      Map<String, String> headers = config.getRequestHeadersMap();
+
+      Header authorizationHeaders = config.getAuthorizationHeaders();
+      headers.putAll(config.getHeadersMap(String.valueOf(authorizationHeaders)));
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/http/sink/batch/HTTPRecordWriter.java]: Although it does not result into NullPointerException here, better to add a null check for authorizationHeader

File: src/main/java/io/cdap/plugin/http/sink/batch/HTTPRecordWriter.java [hunk 30]
```diff
@@ -109,13 +108,11 @@ private void executeHTTPService() throws IOException {
     do {
       HttpURLConnection conn = null;
 
-      Map<String, String> headers = null;
+      Map<String, String> headers = config.getRequestHeadersMap();
+
+      Header authorizationHeaders = config.getAuthorizationHeaders();
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/http/sink/batch/HTTPRecordWriter.java]: Nit : Should be authorizationHeader instead.

**Review by MrRahulSharma — APPROVED** [review]

**Comment by MrRahulSharma** [comment]
LGTM
